src/io.c

```diff
@@ -0,0 +1,2024 @@
+
+        /* put the character on this line                                  */
+        stream->line[stream->pos++] = ch;
+    }
+
+    else {
+        /* position to split                                              */
+        if ((hint = nrLineBreak(stream)) != -1)
+            spos = stream->hints[3 * hint];
+        else
+            spos = 0;
+
+        /* if we are going to split at the end of the line, and we are
+           formatting discard blanks */
+        if (stream->format && spos == stream->pos && ch == ' ') {
+            ;
+        }
+
+        /* full line, acceptable split position */
+        else if (stream->format && spos != 0) {
+
+            /* add character to the line, terminate it */
+            stream->line[stream->pos++] = ch;
+            stream->line[stream->pos++] = '\0';
+
+            /* copy the rest after the best split position to a safe place */
+            for (i = spos; i < stream->pos; i++)
+                str[i - spos] = stream->line[i];
+            str[i - spos] = '\0';
+
+            /* print line up to the best split position */
+            stream->line[spos++] = '\n';
+            stream->line[spos] = '\0';
+            PutLineTo(stream, spos);
+            spos--;
+
+            /* indent for the rest */
+            stream->pos = 0;
+            for (i = 0; i < stream->hints[3 * hint + 2]; i++)
+                stream->line[stream->pos++] = ' ';
+            spos -= stream->hints[3 * hint + 2];
+
+            /* copy the rest onto the next line */
+            for (i = 0; str[i] != '\0'; i++)
+                stream->line[stream->pos++] = str[i];
+            /* recover line break hints for copied rest */
+            for (i = hint + 1; stream->hints[3 * i] != -1; i++) {
+                stream->hints[3 * (i - hint - 1)] =
+                    stream->hints[3 * i] - spos;
+                stream->hints[3 * (i - hint - 1) + 1] =
+                    stream->hints[3 * i + 1];
+                stream->hints[3 * (i - hint - 1) + 2] =
+                    stream->hints[3 * i + 2];
+            }
+            stream->hints[3 * (i - hint - 1)] = -1;
+        }
+
+        /* full line, no split position */
+        else {
+
+            if (stream->format) {
+                /* append a '\',*/
+                stream->line[stream->pos++] = '\\';
+                stream->line[stream->pos++] = '\n';
+            }
+            /* and print the line                                */
+            stream->line[stream->pos] = '\0';
+            PutLineTo(stream, stream->pos);
+
+            /* add the character to the next line */
+            stream->pos = 0;
+            stream->line[stream->pos++] = ch;
+
+            if (stream->format)
+                stream->hints[0] = -1;
+        }
+    }
+}
+
+/****************************************************************************
+ **
+ *F  FuncToggleEcho( )
+ **
+*/
+
+Obj FuncToggleEcho(Obj self)
+{
+    STATE(Input)->echo = 1 - STATE(Input)->echo;
```

Comment:
Future cleanup: change this to the more usual ` STATE(Input)->echo = !STATE(Input)->echo;`